openfl/transport/grpc/director_server.py

```diff
@@ -206,22 +208,29 @@ async def StreamMetrics(self, request, context):  # NOQA:N802
         """Request to stream metrics from the aggregator to frontend."""
         if not self.validate_caller(request, context):
             return
-        # We should probably set a name to the aggregator and verify it here.
-        # Moreover, we may save the experiment name in plan.yaml and retrieve it
-        # during the aggregator initialization
+
         logger.info(f'Request StreamMetrics for {request.experiment_name} experiment has got!')
 
-        async for message in \
-                self.director.stream_metrics(request.experiment_name, request.header.sender):
-            yield message
+        while True:
+            metric_dict = self.director.get_next_metric(
+                experiment_name=request.experiment_name, caller=request.header.sender
+            )
+            if len(metric_dict) == 0:
+                await asyncio.sleep(5)
```

Comment:
It looks like we are bringing internal logic to protocol-transport layer. Why are we waiting several seconds when we got empty dict? Why do empty dict and None lead to different logic?

Reply:
Please refer to the `get_next_metric` method's docstring. The signature of the method is explained there.

Reply:
I mean, these questions are talking about taking out internal logic

Reply:
Changed the get_next_metric method